src/runtime/types.cpp

```diff
@@ -1675,6 +1676,18 @@ static Box* functionCode(Box* self, void*) {
     return codeForFunction(func);
 }
 
+static void functionSetCode(Box* self, Box* v, void*) {
+    assert(self->cls == function_cls);
+
+    if (v == NULL || !PyCode_Check(v))
+        raiseExcHelper(TypeError, "__code__ must be set to a code object");
+
+    BoxedFunction* func = static_cast<BoxedFunction*>(self);
+    BoxedCode* code = static_cast<BoxedCode*>(v);
+    func->f = code->f;
+    func->dependent_ics.invalidateAll();
```

Comment:
We haven't been super vigilant about using dependent_ics.  I took a quick look and didn't see much that could go wrong, but could we have an assert here that `f->internal_callable == NULL`?  callFunc (the default if no internal_callable is set) will use dependent_ics properly but I'm not sure about the different internal_callable implementations.  Also, if you did try to set the code object on a function with an internal_callable set, we'd have to null it out so that we didn't continue to call it.

I think the same kind of thing can apply to tpp_call, if you tried to do something like `MethodType.__call__.func_code = foo`.  CPython won't even allow that, but I think we might, but then it wouldn't have any effect.  Maybe one way to solve this to use something similar to the function `can_change_defaults` flag which will opt-in a function to this extra dynamicism, and then only set it for runtime-created functions.  We could potentially even just reuse (and rename) that same flag for this.